stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -457,66 +457,39 @@ generic_norm1(x) = mapreduce(float ∘ norm, +, x)
 
 # faster computation of norm(x)^2, avoiding overflow for integers
 norm_sqr(x) = norm(x)^2
-norm_sqr(x::Number) = abs2(x)
-norm_sqr(x::Union{T,Complex{T},Rational{T}}) where {T<:Integer} = abs2(float(x))
+norm_sqr(x::T) where {T<:Number} = abs2(promote_type(T,Float64)(x))
 
 function generic_norm2(x)
-    maxabs = normInf(x)
-    (maxabs == 0 || isinf(maxabs)) && return maxabs
-    (v, s) = iterate(x)::Tuple
-    T = typeof(maxabs)
-    if isfinite(length(x)*maxabs*maxabs) && maxabs*maxabs != 0 # Scaling not necessary
-        sum::promote_type(Float64, T) = norm_sqr(v)
-        while true
-            y = iterate(x, s)
-            y === nothing && break
-            (v, s) = y
-            sum += norm_sqr(v)
-        end
-        return convert(T, sqrt(sum))
-    else
-        sum = abs2(norm(v)/maxabs)
-        while true
-            y = iterate(x, s)
-            y === nothing && break
-            (v, s) = y
-            sum += (norm(v)/maxabs)^2
-        end
-        return convert(T, maxabs*sqrt(sum))
+    isempty(x) && return float(norm(zero(eltype(itr))))
+    T = typeof(float(norm(first(x))))
+    sT = promote_type(T, Float64)
+    ans = mapreduce(norm_sqr, +, x)
+    ans in (0, Inf) || return convert(T, sqrt(ans))
+    maxabs = sT(normInf(x))
+    ans = sT(0)
+    for v in x
+        ans += (norm(v)/maxabs)^2
```

Comment:
I can't measure any change to pulling out the division, and multiplying by `invmaxabs`. 

But adding `@simd for v in x` seems to help quite a bit. Is it safe to do so?

Reply:
The second. I don't want to deal with the `invmaxabs` here since we're already in a slow path.

Reply:
FWIW, times for me with `rand(1000)` are 

* `for v in x; out += (norm(v)/maxabs)^2` takes 2.352 μs
* with `@simd` 1.733 μs, 
* both after `normInf(x)` which takes 1.404 μs, same as `maximum`

vs:

* `BLAS.nrm2` takes 1.196 μs
* `mapreduce(norm_sqr, +, x)` takes 229.763 ns

So I guess LinearAlgebra.NRM2_CUTOFF should be higher, currently 32.

But also, why is maximum so slow? Can this be done less carefully here since we don't care about -0.0 and NaN?